### PR TITLE
Modifications for CP_DEPEND, auto classpath generation

### PR DESCRIPTION
**Pull request by wltjr**
@gentoo-java @chewi @monsieurp @fordfrog @austin987 Initial modifications for auto classpath via new variable CP_DEPEND. In most cases the COMMON_DEPEND/CDEPEND can just be renamed to CP_DEPEND. You do need to remove any manual classpath creation/variable setting from the ebuild. This should make it much easier to create and maintain ebuilds. It is a common mistake to have a slot in dep, but use a different slot in the classpath. This prevents such issues, and eliminates need for any variables shared between deps and classpath. This has NO effect unless CP_DEPEND variable is present in an ebuild. Thus should be 100% safe to commit, as it does not change any existing function. If the ebuild does not have CP_DEPEND, this will have no effect.

The function java-pkg_gen-cp may need to be modified code wise. I wanted to avoid using sed, awk, or anything external. Not sure if that effects performance, or uses more resources to do it all in bash. Feel free to tweak/recode the internal of java-pkg_gen-cp function. It is not intended to handle complex deps, no || or other. That is not used to commonly in shared depends now. I do not think that will be a problem going forward. If it is the function will need to be modified to handle more complex depend scenarios. But I do not believe we have any || in any classpath variables, which means the same for deps, so usage of such is not to likely.

### CODE REVIEW

File: eclass/java-utils-2.eclass [hunk 1]
```diff
@@ -2886,3 +2886,27 @@ java-pkg_clean() {
...
+# @CODE
+# Parameters:
+# $1 - classpath variable either EANT_GENTOO_CLASSPATH or JAVA_GENTOO_CLASSPATH
+# @CODE
+java-pkg_gen-cp() {
+	debug-print-function ${FUNCNAME} "${@}"
+
+	if [[ ${CP_DEPEND} ]]; then
+		local cp="${!1}"
+		for p in ${CP_DEPEND}; do
```
Comment (monsieurp) [inline eclass/java-utils-2.eclass]: Make `p` local as well pls.

**Comment by wltjr** [comment]
Made p local

**Comment by chewi** [comment]
I was initially sceptical about whether this was a good idea but it's simpler than I expected and uses a dedicated variable so you can still do certain dependencies the old-fashioned way if you need to. I think the string manipulation needs some work so I'll take a look at that very soon. Only other suggestion is that we join the atoms with `,` rather than a space. As things stand, java-pkg_getjars is called in a loop from eant and you can reduce the number of calls by using `,`. I should fix this but using `,` is also handy if you need to call java-config manually, which I sometimes do when building/running tests.

**Comment by wltjr** [comment]
Usage of commas breaks its function.

!!! ERROR: Package  was not found!

I believe it is using java-pkg_getjar not the others. Both java-pkg_jar-from and java-pkg_getjars have for pkg in ${target_pkg//,/ }; do

Usage of the space is correct and comma will cause breakage. I am not changing how the CLASSPATH is used/parsed by other eclasses, ant and simple. They seem to be calling java-pkg_getjar, so the comma is not handled properly. Though from looking at it, if we ommit commas in the first place we do not have to do //,/ to variables. That might have been some transition thing that stuck longer than it should have.

**Comment by wltjr** [comment]
I agree the string manipulation was not ideal. Bash does not seem to allow nested substitution and short of using sed or awk. Best way I could come up with was multiple passes. It is ugly, but not sure if calling sed/awk would be less efficient, albeit more elegant code wise. If efficiency matters over elegance. I have no tests showing one way is more efficient than the other. Just speculation doing it all within bash is faster than invoking external and using more resources.

**Comment by wltjr** [comment]
The eclass is calling getjars, not sure why commas make it fail, have to debug that.

**Comment by chewi** [comment]
The `for pkg in ${target_pkg//,/ }; do` bit is only around the call to `java-pkg_ensure-dep`. I believe the error is from setting cp with a leading `,`. Strip that off and it'll probably work.

**Comment by wltjr** [comment]
It is what I figured, fails on first run due to the cp variable not being set to you end up with ,${cp}, where using a space does not harm. Have to modify code to have an initialize case, and then another for the remainder. Or go with spaces, which I do not believe will be harmful and require less code. Either way, I will code up alternative, and be ready to push if you elect to go that route. But if spaces will not cause breakage, may just stick with them. Since it seems commas are just later converted to spaces anyway.

Only issue I could see is where other stuff is added with commas. But then again since the commas get converted to spaces. Does it matter?

**Comment by wltjr** [comment]
The difference is this.

`                        if [[ -z ${cp} ]]; then
                                cp="${p#*/}"
                        else    
                                cp="${cp},${p#*/}"
                        fi`

 Which then adds a conditional on every run, when it is only needed on the first. Could do it outside the loop for the first, but then I think the patter for will get duplicated and create more places to maintain the code.

**Comment by wltjr** [comment]
Or could do this, but not sure it is possible to have the conditional be outside the loop so will always be run if that matters.

`if [[ ! -z ${cp} ]]; then
                                cp="${cp},"
                        fi 
                        cp="${cp}${p#*/}"`

Can also use += vs passing cp.

**Comment by wltjr** [comment]
This
`if [[ ! -z ${cp} ]]; then
                                cp+=","
                        fi
                        cp+="${p#*/}"`

vs just
cp+=" ${p#*/}"``

Then there is the other stuff, which if there is another way I am open to that. But not sure about another way to append the comma. If we should do that or stick with spaces. It is minimal overhead, but the commas just get converted to spaces. Which may be senseless to use commas for that reason.

**Comment by chewi** [comment]
I've now boiled it down to this.

```bash
	local cp p

	for p in ${CP_DEPEND}; do
		p=${p/-[0-9]*:/-}
		cp+=${cp:+,}${p#*/}
	done

	if [[ ${cp} ]]; then
		cp=${cp//:/-}
		declare -g ${1}+=${!1:+,}${cp//-0/}
	fi
```

The commas are sorted out but the string manipulation is far too naive. What if the version string was something like this?

`>=dev-java/foo-bar-1.2.3_alpha4-r5:6xyz=[baz]`

I'll see if I can crack it.

**Comment by wltjr** [comment]
I am not sure how advanced the string format for most java deps is now. It is open to change and improvement to prevent potential breakage.

**Comment by chewi** [comment]
I have merged this but made an additional commit to replace the function with my own version. It should be bulletproof now.

**Comment by wltjr** [comment]
Not surprised, most things have to be "Chewi's" way vs collaboration. Guess there is an I in team, or at least is in Chewi. 

I see you added the comma back despite me showing it gets filtered out internally so is pointless.

Also did you test this?
```
JAVA_GENTOO_CLASSPATH=commons-cli-1,asm-3
 * java-pkg_ensure-dep: The ebuild is attempting to use commons-cli-1, which is not declared with a SLOT in DEPEND.
 * java-pkg_ensure-dep: The ebuild is attempting to use asm-3, which is not declared with a SLOT in DEPEND.
 * Compiling ...
```

The version I proposed does not generate such output. The modifications caused some other eclass issues. Not so bulletproof, vest already has a hole in it with blood coming out... Though granted not breakage as in causing anything to fail.

**Comment by chewi** [comment]
> I see you added the comma back despite me showing it gets filtered out internally so is pointless.

What you showed was the wrong piece of code. That snippet was from `java-pkg_jar-from`, which is not relevant here. `java-pkg_getjars` only accepts commas and does a single call with that comma-separated list to `java-config`. It only converts to spaces when calling `java-pkg_ensure-dep` and `java-pkg_record-jar_`.

The important functions are `eant` and `java-pkg-simple_src_compile`, which loop over `EANT_GENTOO_CLASSPATH` and `JAVA_GENTOO_CLASSPATH` respectively. If they have spaces, `java-pkg_getjars` and by extension `java-config` are called many times. If they have commas, they are called only once.

> Also did you test this?

It looks like you forgot to include `CP_DEPEND` in `DEPEND`. Your version did not do this either. If you still think it's wrong, post a full example so I can replicate it.

**Comment by wltjr** [comment]
Nope here is the entire java-pkg_jar-from. Do you not see the two```${pkgs//,/ };``` below in the function you say they do not exist in?

I am familiar with how java-config operates... It has been working for some time in my overlay with spaces instead of commas that get removed...

```bash
java-pkg_getjars() {
        debug-print-function ${FUNCNAME} $*

        local build_only=""
        local deep=""

        [[ "${EBUILD_PHASE}" == "test" ]] && build_only="build"

        while [[ "${1}" == --* ]]; do
                if [[ "${1}" = "--build-only" ]]; then
                        build_only="build"
                elif [[ "${1}" = "--with-dependencies" ]]; then
                        deep="--with-dependencies"
                else
                        die "java-pkg_jar-from called with unknown parameter: ${1}"
                fi
                shift
        done

        [[ ${#} -ne 1 ]] && die "${FUNCNAME} takes only one argument besides --*"


        local pkgs="${1}"

        if [[ "${EAPI}" == "1" ]]; then
                pkgs="${pkgs//:/-}"
        fi

        jars="$(java-config ${deep} --classpath=${pkgs})"
        [[ $? != 0 ]] && die "java-config --classpath=${pkgs} failed"
        debug-print "${pkgs}:${jars}"

        for pkg in ${pkgs//,/ }; do
                java-pkg_ensure-dep "${build_only}" "${pkg}"
        done

        for pkg in ${pkgs//,/ }; do
                if [[ -z "${build_only}" ]]; then
                        java-pkg_record-jar_ "${pkg}"
                else
                        java-pkg_record-jar_ --build-only "${pkg}"
                fi
        done

        echo "${jars}"
}


```

**Comment by wltjr** [comment]
The java-config argument makes sense for the commas vs spaces, if it does cause multiple runs. I have not observed such so will have to go test and see if that is the case. I am not sure if jem handles things the same way or not.

**Comment by chewi** [comment]
Those two `for pkg in ${pkgs//,/ }; do` loops could be merged into one.

**Pull request closed without merging**